src/animated/Controller.ts

```diff
@@ -281,129 +296,111 @@ class Controller<DS extends object = any> {
+          const toValue = target
+            ? target.animations[name].node
+            : (isInterpolated && 1) || value
+
+          // Animatable strings use interpolation
+          if (isInterpolated) {
+            let parent: AnimatedValue
+            const output = [fromValue, goalValue]
+            if (node instanceof AnimatedInterpolation) {
+              parent = animatedValues[0]
 
-          toValues = toArray(target ? toValue.getPayload() : toValue)
-          animatedValues = toArray(parent.getPayload())
-          if (reset && !isInterpolation) parent.setValue(fromValue, false)
-
-          // Reset animated values
-          animatedValues.forEach(value => {
-            value.startPosition = value.value
-            value.lastPosition = value.value
-            value.lastVelocity = isActive ? value.lastVelocity : undefined
-            value.lastTime = isActive ? value.lastTime : undefined
-            value.startTime = now()
-            value.done = false
-            value.animatedStyles.clear()
-          })
-
-          // Set immediate values
-          if (callProp(immediate, name)) {
-            parent.setValue(isInterpolation ? toValue : value, false)
+              if (!reset) output[0] = node.calc(parent.value)
+              node.updateConfig({ output })
+
+              parent.reset(isActive)
+              parent.setValue(0, false)
+            } else {
+              parent = new AnimatedValue(0)
+              node = parent.interpolate({ output })
+            }
+            if (isImmediate) {
+              parent.setValue(toValue, false)
+            }
+          } else {
+            // Convert values into Animated nodes (reusing nodes whenever possible)
+            if (isArray) {
+              if (node instanceof AnimatedValueArray) {
+                animatedValues.forEach(node => node.reset(isActive))
+              } else {
+                node = new AnimatedValueArray(fromValue)
+              }
+            } else {
+              if (node instanceof AnimatedValue) {
+                node.reset(isActive)
+              } else {
+                node = new AnimatedValue(fromValue)
+              }
+            }
+            if (reset || isImmediate) {
+              node.setValue(reset ? fromValue : goalValue, false)
+            }
           }
 
           changed = true
           acc[name] = {
             ...entry,
             name,
-            parent,
-            interpolation,
-            animatedValues,
-            toValues,
-            previous: newValue,
-            config: toConfig,
-            fromValues: toArray(parent.getValue()),
-            immediate: callProp(immediate, name),
+            node,
+            immediate: isImmediate,
+            goalValue,
+            toValues: toArray(target ? toValue.getPayload() : toValue),
+            fromValues: toArray(node.getValue()),
+            animatedValues: toArray(node.getPayload() as any),
```

Comment:
The `toValues` and `animatedValues` are now calculated directly on the animation object.